pkg/backup/util/util.go

```diff
@@ -123,83 +134,122 @@ func GenerateGcsCertEnvVar(secret *corev1.Secret, gcs *v1alpha1.GcsStorageProvid
 			Value: gcs.StorageClass,
 		},
 		{
-			Name:  "GCS_SERVICE_ACCOUNT_JSON_KEY",
-			Value: string(secret.Data[constants.GcsCredentialsKey]),
+			Name: "GCS_SERVICE_ACCOUNT_JSON_KEY",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{Name: gcs.SecretName},
+					Key:                  constants.GcsCredentialsKey,
+				},
+			},
 		},
 	}
 	return envVars, "", nil
 }
 
 // GenerateStorageCertEnv generate the env info in order to access backend backup storage
-func GenerateStorageCertEnv(backup *v1alpha1.Backup, secretLister corelisters.SecretLister) ([]corev1.EnvVar, string, error) {
-	ns := backup.GetNamespace()
-	name := backup.GetName()
-
+func GenerateStorageCertEnv(ns string, storageType v1alpha1.BackupStorageType, provider v1alpha1.StorageProvider, secretLister corelisters.SecretLister) ([]corev1.EnvVar, string, error) {
 	var certEnv []corev1.EnvVar
 	var reason string
 
-	switch backup.Spec.StorageType {
+	switch storageType {
 	case v1alpha1.BackupStorageTypeS3:
-		s3SecretName := backup.Spec.S3.SecretName
+		if provider.S3 == nil {
+			return certEnv, "S3ConfigIsEmpty", errors.New("s3 config is empty")
+		}
+		s3SecretName := provider.S3.SecretName
 		secret, err := secretLister.Secrets(ns).Get(s3SecretName)
 		if err != nil {
-			err := fmt.Errorf("backup %s/%s get s3 secret %s failed, err: %v", ns, name, s3SecretName, err)
+			err := fmt.Errorf("get s3 secret %s/%s failed, err: %v", ns, s3SecretName, err)
 			return certEnv, "GetS3SecretFailed", err
 		}
 
 		keyStr, exist := CheckAllKeysExistInSecret(secret, constants.S3AccessKey, constants.S3SecretKey)
 		if !exist {
-			err := fmt.Errorf("backup %s/%s, The s3 secret %s missing some keys %s", ns, name, s3SecretName, keyStr)
+			err := fmt.Errorf("s3 secret %s/%s missing some keys %s", ns, s3SecretName, keyStr)
 			return certEnv, "s3KeyNotExist", err
 		}
 
-		certEnv, reason, err = GenerateS3CertEnvVar(secret, backup.Spec.S3.DeepCopy())
+		certEnv, reason, err = GenerateS3CertEnvVar(provider.S3.DeepCopy())
 		if err != nil {
 			return certEnv, reason, err
 		}
 	case v1alpha1.BackupStorageTypeGcs:
-		gcsSecretName := backup.Spec.Gcs.SecretName
+		if provider.Gcs == nil {
+			return certEnv, "GcsConfigIsEmpty", errors.New("gcs config is empty")
+		}
+		gcsSecretName := provider.Gcs.SecretName
 		secret, err := secretLister.Secrets(ns).Get(gcsSecretName)
 		if err != nil {
-			err := fmt.Errorf("backup %s/%s get gcs secret %s failed, err: %v", ns, name, gcsSecretName, err)
+			err := fmt.Errorf("get gcs secret %s/%s failed, err: %v", ns, gcsSecretName, err)
 			return certEnv, "GetGcsSecretFailed", err
 		}
 
 		keyStr, exist := CheckAllKeysExistInSecret(secret, constants.GcsCredentialsKey)
```

Comment:
This is already checked in `GenerateGcsCertEnvVar`.

Reply:
`GenerateGcsCertEnvVar` does not check whether the secret exists and contains the specified key.